src/platform/PICO/adc_pico.c

```diff
@@ -0,0 +1,260 @@
+#define PICO_ADC_FIFO_SIZE              8
+#define PICO_ADC_CHANNEL_COUNT          9
+#define PICO_ADC_INTERNAL_TEMP_CHANNEL  8
+#else
+#error "ADC not properly defined, perhaps incorrect PICO target?"
+#endif
+
+#if defined(USE_ADC_INTERNAL)
+#warning "PICO: Internal temp/Vref are used for ADC padding for DMA only - not intended for actual use"
+#endif
+
+#if ADC_SOURCE_COUNT > PICO_ADC_MAX_CHANNELS
+#warning "PICO currently only supports maximum of 4 ADC channels"
+#endif
+
+typedef struct adcOperatingConfig_s {
+    ioTag_t tag;
+    uint8_t channel;  // hardware channel number for this input.
+    uint8_t dmaIndex;    // location in ADC values for this input.
+    bool enabled;
+} adcOperatingConfig_t;
+
+static adcOperatingConfig_t adcOperatingConfig[ADC_SOURCE_COUNT];
+static volatile uint16_t adcValues[PICO_ADC_CHANNEL_COUNT];
+
+static int adcChannelByPin(const int pin)
+{
+#ifdef RP2350A
+    if (pin >= 26 && pin <= 29) {
+        return pin - 26;
+    }
+#elif defined(RP2350B)
+    if (pin >= 40 && pin <= 47) {
+        return pin - 40;
+    }
+#else
+    UNUSED(pin);
+#endif
+    return -1;
+}
+
+static inline bool is_adc_busy(void)
+{
+    // The READY bit is 0 while a conversion is in progress.
+    return !(adc_hw->cs & ADC_CS_READY_BITS);
+}
+
+void adcInit(const adcConfig_t *config)
+{
+    adc_init();
+
+    memset(adcOperatingConfig, 0, sizeof(adcOperatingConfig));
+
+    // Set channels (ioTags) for enabled ADC inputs
+    if (config->vbat.enabled) {
+        adcOperatingConfig[ADC_BATTERY].tag = config->vbat.ioTag;
+    }
+
+    if (config->rssi.enabled) {
+        adcOperatingConfig[ADC_RSSI].tag = config->rssi.ioTag;
+    }
+
+    if (config->external1.enabled) {
+        adcOperatingConfig[ADC_EXTERNAL1].tag = config->external1.ioTag;
+    }
+
+    if (config->current.enabled) {
+        adcOperatingConfig[ADC_CURRENT].tag = config->current.ioTag;
+    }
+
+    uint32_t mask = 0u;
+
+    // loop over all possible channels and build the adcOperatingConfig to represent
+    // the set of enabled channels
+    for (int i = 0; i < ADC_SOURCE_COUNT; i++) {
+        if (!adcOperatingConfig[i].tag || i > PICO_ADC_MAX_CHANNELS) {
+            continue;
+        }
+
+        const int pin = DEFIO_TAG_PIN(adcOperatingConfig[i].tag);
+        const int channel = adcChannelByPin(pin);
+        if (channel >= 0) {
+            adcOperatingConfig[i].channel = channel;
+            adc_gpio_init(pin);
+            adcOperatingConfig[i].enabled = true;
+        }
+        mask |= (1u << channel);
+    }
+
+    // Map each enabled source to its DMA slot by channel rank (ascending channel order).
+    for (int i = 0; i < ADC_SOURCE_COUNT; i++) {
+        if (!adcOperatingConfig[i].enabled) {
+            continue;
+        }
+        const uint8_t ch = adcOperatingConfig[i].channel;
+        adcOperatingConfig[i].dmaIndex = (uint8_t)popcount(mask & ((1u << ch) - 1u));
+    }
+
+    const uint8_t channelCount = popcount(mask);
```

Comment:
```suggestion

    // loop over all possible channels and build the adcOperatingConfig to represent
    // the set of enabled channels
    int channelsEnabled = 0;
    uint32_t mask = 0u;
    for (int i = 0; i < ADC_SOURCE_COUNT && channelsEnabled < PICO_ADC_MAX_CHANNELS; i++) {
        if (!adcOperatingConfig[i].tag) {
            continue;
        }

        const int pin = DEFIO_TAG_PIN(adcOperatingConfig[i].tag);
        const int channel = adcChannelByPin(pin);
        if (channel < 0) {
            continue;
        }
        adc_gpio_init(pin);
        adcOperatingConfig[i].enabled = true;
        adcOperatingConfig[i].channel = channel;
        adcOperatingConfig[i].dmaIndex = channelsEnabled;
        mask |= (1u << channel);
        channelsEnabled++;
    }
    const uint8_t channelCount = channelsEnabled;
```

(maybe use channelsEnabled in rest of function)

Reply:
The order of the mask - dictated by the channel number (not the enum position) is what drives the dmaIndex - not the enum position. Therefore if the pins are not in order the channels will not be in order, and so one needs to evaluate the mask - not the enum.